### PR TITLE
Add python packager

### PR DESCRIPTION
**Pull request by stealthycoin**
Add low level python packager

Port over the python dependency builder from Chalice. Also added a small
wrapper class to adapt the interface to something more like what the
Builder Action will use. The interface for the low level will stay the
same, the wrapper PythonPipDependencyBuilder class can be modified as
needed to match the expected interface as it evolves.

I made a few other changes as well to get the tests running, make things importable, etc.

### CODE REVIEW

File: aws_lambda_builders/actions/python_pip/package.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
+import sys
+import hashlib
+import inspect
```
Comment (garnaat) [inline aws_lambda_builders/actions/python_pip/package.py]: hash lib and inspect are not used in the module.  I know it's nitpicky but I'm a nitpicker 8^)

File: aws_lambda_builders/actions/python_pip/package.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+import re
+import subprocess
+from email.parser import FeedParser
+from email.message import Message
+from zipfile import ZipFile
```
Comment (garnaat) [inline aws_lambda_builders/actions/python_pip/package.py]: ditto for Message and ZipFile

File: aws_lambda_builders/workflows/python_pip/actions.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from aws_lambda_builders.actions import BaseAction
+from .packager import PythonPipDependencyBuilder
+
+class PythonPipBuildAction(BaseAction):
+
+    NAME = 'PythonPipBuildAction'
+
+    def execute(self):
+        # TODO still needs to be worked out how these are plumbed through from the workflow
```
[the remark this answers is not in the document]
Comment (sanathkr) [reply aws_lambda_builders/workflows/python_pip/actions.py]: I fixed this myself and pushed to your branch

File: aws_lambda_builders/workflows/python_pip/workflow.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from aws_lambda_builders.workflows import BaseWorkflow
+from aws_lambda_builders.actions import CopySourceAction
+
+from .actions import PythonPipBuildAction
+
+
+class PythonPipWorkflow(BaseWorkflow):
+    def __init__(self, *args, **kwargs):
+        super(PythonPipWorkflow, self).__init__(*args, **kwargs)
+        # TODO if the parent class just has a self.ACTIONS  that it uses we can
```
[the remark this answers is not in the document]
Comment (sanathkr) [reply aws_lambda_builders/workflows/python_pip/workflow.py]: I fixed this myself and pushed to your branch

File: aws_lambda_builders/workflows/python_pip/packager.py [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+        :param target_directory: The directory to build all dependencies into.
+            This directory should be on the PYTHON_PATH of whichever process
+            wants to use thse dependencies.
+
+        :raises MissingDependencyError: This exception is raised if one or more
+            packages could not be installed. The complete list of missing
+            packages is included in the error object's ``missing`` property.
+        """
+        if self._has_at_least_one_package(requirements_filepath):
+            with self._osutils.tempdir() as tempdir:
```
[the remark this answers is not in the document]
Comment (sanathkr) [reply aws_lambda_builders/workflows/python_pip/packager.py]: Created #12 to address this

File: aws_lambda_builders/workflows/python_pip/packager.py [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+        elif sdist_path.endswith(('.tar.gz', '.tar.bz2')):
+            self._osutils.extract_tarfile(sdist_path, unpack_dir)
+        else:
+            raise InvalidSourceDistributionNameError(sdist_path)
+        # There should only be one directory unpacked.
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        return self._osutils.joinpath(unpack_dir, contents[0])
+
+    def get_package_name_and_version(self, sdist_path):
+        with self._osutils.tempdir() as tempdir:
```
[the remark this answers is not in the document]
Comment (sanathkr) [reply aws_lambda_builders/workflows/python_pip/packager.py]: #12

File: aws_lambda_builders/workflows/python_pip/packager.py [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+            and errors. If injected by the caller, it can be used to monitor
+            the status of the build process or forward this information
+            elsewhere.
+
+        :type config: :class:`lambda_builders.utils.Config` or None
+        :param config: To be determined. This is an optional config object
+            we can extend at a later date to add more options to how pip is
+            called.
+        """
+        # TODO: The DependencyBuilder makes the assumption that it is running
```
[the remark this answers is not in the document]
Comment (sanathkr) [reply aws_lambda_builders/workflows/python_pip/packager.py]: Created #13 to investigate the virtualenv story

**Pull request merged**